### PR TITLE
feat(panel): Clean up promises

### PR DESCRIPTION
**Pull request by KarenParker**
@ErinCoughlan @gmoothart @jelbourn @DerekLouie @ThomasBurleson  Please review.

### CODE REVIEW

File: src/components/panel/panel.js [hunk 1]
```diff
@@ -861,20 +837,15 @@ MdPanelRef.prototype.open = function() {
...
-        return self.detach();
-      }, this._$q.reject)
-      .then(function() {
-        // TODO(ErinCoughlan) - Add destroy. This will make the code here
-        // different than just calling this.detach().
-        return self._$q.resolve(self);
-      }, this._$q.reject);
-
-  return closePromise;
+  return this._$q(function(resolve, reject) {
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Careful with the cluttering of rejects. I contend that this is a better structure:

```js
MdPanelRef.prototype.close = function() {
 var self = this; 
 
return this._$q(function(resolve, reject) {
    var done = angular.bind(self,resolve, self),
        detach = angular.bind(self, self.detach);
        
    self.hide()
        .then( detach )
        .then( done )
        .catch( reject );
 });
};
```

> ^ is untested but I believe the logic is valid and more clear.

File: src/components/panel/panel.js [hunk 2]
```diff
@@ -1003,31 +968,35 @@ MdPanelRef.prototype.show = function() {
...
+    return this._$q(function(resolve, reject) {
+      reject('Panel does not exist yet. Call open() or attach().');
+    });
   }
 
-  if (this._panelContainer.hasClass(MD_PANEL_HIDDEN) && this._hidePromise) {
-    return this._hidePromise;
+  var self = this;
+  if (this._panelContainer.hasClass(MD_PANEL_HIDDEN)) {
+    return this._$q(function(resolve) {
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Needless verbose... use `return this._$q.when(self)`;

File: src/components/panel/panel.js [hunk 3]
```diff
@@ -1361,22 +1342,25 @@ MdPanelRef.prototype._animateOpen = function() {
  * @private
  */
 MdPanelRef.prototype._animateClose = function() {
+  var self = this;
   var animationConfig = this._config['animation'];
   if (!animationConfig) {
     this.removeClass('md-panel-is-showing');
     this.removeClass('_md-panel-shown');
-    return this._$q.resolve();
+    return this._$q(function(resolve) {
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: `return this._$q.when(self);`

File: src/components/panel/panel.js [hunk 4]
```diff
@@ -2044,11 +2028,22 @@ MdPanelAnimation.prototype.animateOpen = function(panelEl, animator) {
...
-    return this._reverseAnimation();
-  }
-  return $q.reject('No panel close animation. ' +
-      'Have you called MdPanelAnimation.animateOpen()?');
+  var self = this;
+  return $q(function(resolve, reject) {
+    debugger;
+    if (self._reverseAnimation) {
+      self._reverseAnimation().then(function() {
+        debugger;
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Remove `debugger` statements

File: src/components/panel/panel.js [hunk 5]
```diff
@@ -2044,11 +2028,22 @@ MdPanelAnimation.prototype.animateOpen = function(panelEl, animator) {
...
+        resolve(self);
+      }, function() {
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Usually you specify the error argument and reject with that:

```js
}, function(error) { 
    reject(errror); 
});

File: src/components/panel/panel.js [hunk 6]
```diff
@@ -1343,15 +1316,23 @@ MdPanelRef.prototype._configureTrapFocus = function() {
...
+    return this._$q(function(resolve) {
+      resolve(self);
+    });
   }
 
-  return animationConfig.animateOpen(this._panelEl,
-      this._$mdUtil.dom.animator);
+  return this._$q(function(resolve, reject) {
+    animationConfig.animateOpen(self._panelEl, self._$mdUtil.dom.animator)
+        .then(function() {
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]:
```js
.then(function() {
   resolve(self);
}, 
``` 
is used in many places. Recommend you do this:

````
var respondWith = function(callback, value) {
   return angular.bind(null, callback, value);
}
```
then you can reuse this:

```js
var animator = self._$mdUtil.dom.animator;

return this._$q(function(resolve, reject) {
   animationConfig
        .animateOpen(self._panelEl, animator)
        .then( respondWith(resolve, self) )
        .catch( respondWith(reject, self) );
```

File: src/components/panel/panel.spec.js [hunk 7]
```diff
@@ -136,67 +138,69 @@ describe('$mdPanel', function() {
...
-    it('should reject on create when opening', function () {
+    it('should reject on create when opening', function() {
       var openRejected = false;
 
       panelRef._createPanel = function() {
         return panelRef._$q.reject();
       };
 
-      panelRef.open().then(function() {}, function () {
+      panelRef.open().then(function() {
```
Comment (DevVersion) [inline src/components/panel/panel.spec.js]: Using those `noop` functions is not good here.

You're able to use `panelRef.open().catch(errorCallback)`, which is a shorthand for 
```js
promise.then(null, errorCallback)
```

This can be used in the same lines below.

File: src/components/panel/panel.js (hunk 1 above)
Comment (KarenParker) [inline src/components/panel/panel.js]: This will work in some places, but others will need to change the thens differently. According to Erin, we can't guarantee the order the thens are executed if there are more than one.

Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Not true. Each `then( )` returns a new promise; so the chaining sequence is guaranteed.

**Comment by ThomasBurleson** [comment]
@KarenParker  - can you rebase with `master` and squash the 48 commits ?

**Comment by KarenParker** [comment]
Erin broke all of feat/panel. I have it rebased and squashed correctly. All comments addressed and all tests passing.

File: src/components/panel/panel.js [hunk 8]
```diff
@@ -885,42 +858,38 @@ MdPanelRef.prototype.close = function() {
...
+    var done = self._done(resolve, self);
+
+    self._$q.all([
+      self._createBackdrop(),
+      self._createPanel().then(function() {
+        self.isAttached = true;
+        self._addEventListeners();
+      }, reject)
+    ]).then(done)
+        .catch(reject);
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: Do you expect `done()` to reject or throw an exception? 
* If yes, the `.catch(reject)` is good. 
* If no, then you can do `.then( done, reject )`

File: src/components/panel/panel.js [hunk 9]
```diff
@@ -1369,18 +1341,53 @@ MdPanelRef.prototype._animateClose = function() {
...
+ */
+MdPanelRef.prototype._simpleBind = function(callback, self) {
+  return function(value) {
+    return callback.apply(self, value);
+  };
+};
+
+
+/**
+ * //Faster, more basic than angular.bind
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: nit: Strange comment format

**Comment by KarenParker** [comment]
Comments addressed.

File: src/components/panel/panel.spec.js [hunk 10]
```diff
@@ -243,43 +235,39 @@ describe('$mdPanel', function() {
...
       expect(panelRef.isAttached).toEqual(true);
     });
 
-    it('should resolve on animate failure when closing', function () {
+    it('should resolve on animate failure when closing', function() {
       var closeResolved = false;
 
       panelRef.open();
-      $rootScope.$apply();
+      flushPanel();
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: How about using `openPanel()` instead, which is these two lines?

File: src/components/panel/panel.js [hunk 11]
```diff
@@ -1369,18 +1341,53 @@ MdPanelRef.prototype._animateClose = function() {
   if (!animationConfig) {
     this.removeClass('md-panel-is-showing');
     this.removeClass('_md-panel-shown');
-    return this._$q.resolve();
+    return this._$q.when(this);
   }
 
   var self = this;
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: I am a huge advocate of code clarity; try this:

```js
var self = this,
    animator = self._$mdUtil.dom.animator;

return this._$q(function(resolve) {
    var done = function(response) {
            self.removeClass('md-panel-is-showing');
            resolve(self, response);
            return self;
        },
        warnAndClose = function(error) {
            self._$log.warn('MdPanel Animations failed. Hiding panel without animating.');
            return done(error);
        };

    animationConfig
        .animateClose(self._panelEl, animator)
        .then( done )
        .catch( warnAndClose );
});
```

**Comment by KarenParker** [comment]
Done.

**Comment by ThomasBurleson** [comment]
Nice work @KarenParker. Thanks for being so responsive. 👍

File: src/components/panel/panel.js [hunk 12]
```diff
@@ -961,37 +929,29 @@ MdPanelRef.prototype.detach = function() {
...
-      }, self._$q.reject)
-  ]);
-
-  return this._showPromise;
+  return this._$q(function(resolve, reject) {
+    var done = self._done(resolve, self);
+
+    self._$q.all([
+      self._backdropRef ? self._backdropRef.show() : self,
+      animatePromise().then(function() { self._focusOnOpen(); })
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Doesn't this `then` also need a reject? (I think this will break in Angular 1.6)

File: src/components/panel/panel.js [hunk 13]
```diff
@@ -1003,31 +963,24 @@ MdPanelRef.prototype.show = function() {
...
-    }, self._$q.reject)
-  ]);
-
-  return this._hidePromise;
+  return this._$q(function(resolve, reject) {
+    var done = self._done(resolve, self);
+
+    self._$q.all([
+      self._backdropRef ? self._backdropRef.hide() : self,
+      self._animateClose().then(function() { self.addClass(MD_PANEL_HIDDEN); })
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Ditto here with the reject

File: src/components/panel/panel.js [hunk 14]
```diff
@@ -1175,10 +1130,9 @@ MdPanelRef.prototype._focusOnOpen = function() {
     // Wait a digest to guarantee md-autofocus has finished adding the class
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Update this comment? We are waiting for angular to finish rendering the template now rather than a digest.

File: src/components/panel/panel.js [hunk 15]
```diff
@@ -1369,18 +1337,53 @@ MdPanelRef.prototype._animateClose = function() {
...
+        .then(done, warnAndClose);
   });
 };
 
+
+/**
+ * Faster, more basic than angular.bind
+ * http://jsperf.com/angular-bind-vs-custom-vs-native
+ * @param {function} callback
+ * @param {Object} self
```
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: nit: `!Object`

File: src/components/panel/panel.spec.js [hunk 16]
```diff
@@ -27,12 +27,14 @@ describe('$mdPanel', function() {
...
     $mdConstant = $injector.get('$mdConstant');
     $mdUtil = $injector.get('$mdUtil');
+    $animate = $injector.get('$animate');
   };
 
   beforeEach(function() {
     module('material.components.panel', 'ngSanitize');
 
     inject(injectLocals);
+    $animate.enabled(false);
```
Comment (ErinCoughlan) [inline src/components/panel/panel.spec.js]: Re-enable in an afterEach so we don't break others.

**Comment by ErinCoughlan** [comment]
LGTM after comments are addressed.
